### DISCUSSION TITLE
Fetch annotations with annotated element

### DISCUSSION THREAD

**Discussion by abhisarsaraswat** [pymupdf/PyMuPDF#1163] [Looking for help] [answered]
Hey! I am doing decomposition of PDF document. I've done decomposition of the PDF successfully but, Now I want to fetch all annotations mapped with the elements on which they are annotated on. 

For eg: 

![Screenshot from 2021-07-23 12-34-43](https://user-images.githubusercontent.com/42264203/126748856-cdefc946-d6ae-47a7-a171-a72702cfb5ca.png)

In this example, I've a paragraph on which annotations "Paragraph" is annotated. I want to fetch "Paragraph" annotation with the whole paragraph mapped with it.

I've tried " .annots() " But i didn't got the desired output:

#code:
```
import fitz
document = 'Segment_Data_Format.pdf'
doc = fitz.open(document)
for page in doc:
    for annot in page.annots():  # loop thru freetext annots
        text = annot.info
        print(text)
```

#output:
```
{'content': 'Paragraph', 'name': '', 'title': '', 'creationDate': '', 'modDate': '', 'subject': '', 'id': ''}
```

Thanks in Advance,

[Segment_Data_Format.pdf](https://github.com/pymupdf/PyMuPDF/files/6867233/Segment_Data_Format.pdf)

**Comment by JorjMcKie** [comment]
Don't understand yet:

>  But i didn't got the desired output:

What would be the desired output? The page text covered by the annot rectangle?

**Reply by abhisarsaraswat** [reply]
I'm sorry for not elaborating, 
I want a output like this.
```
{
    "text": "Lorem Ipsum is simply dummy text of the printing and typesetting industry.Lorem Ipsum has been the industry 's standard dummy text ever since the 1500 s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged.It was popularised in the 1960 s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum."
    "annotation": "Paragraph"
}
```

If there is any annotation marked on the text, i want to populate it with text. Thanks for the help!

**Comment by JorjMcKie** [comment]
Would this be it:
```python
>>> for a in page.annots():
	print("Annot: %s, text: '%s'" % (a.info["content"], page.get_textbox(a.rect)))

	
Annot: Heading 1, text: ''
Annot: Heading 2, text: ''
Annot: Paragraph, text: '1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also
the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing 
Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum. '
Annot: Reference Link, text: ''
Annot: Heading 2, text: ''
Annot: Paragraph, text: 'quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum 
dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. '
Annot: Heading 3, text: ''
Annot: Paragraph, text: 'inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit, 
sed quia consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt. Neque porro quisquam est, qui dolorem ipsum quia dolor sit amet, 
consectetur, adipisci velit, sed quia non numquam eius modi tempora incidunt ut labore et dolore magnam aliquam quaerat voluptatem. Ut enim ad 
minima veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem vel eum iure 
reprehenderit qui in ea voluptate velit esse quam nihil molestiae consequatur, vel illum qui dolorem eum fugiat quo voluptas nulla pariatur? '
Annot: Heading 4, text: ''
Annot: Paragraph, text: 'system, and expound the actual teachings of the great explorer of the truth, the master-builder of human happiness. No one rejects, dislikes, or avoids 
pleasure itself, because it is pleasure, but because those who do not know how to pursue pleasure rationally encounter consequences that are 
extremely painful. Nor again is there anyone who loves or pursues or desires to obtain pain of itself, because it is pain, but because occasionally 
circumstances occur in which toil and pain can procure him some great pleasure. To take a trivial example, which of us ever undertakes laborious 
physical exercise, except to obtain some advantage from it? But who has any right to find fault with a man who chooses to enjoy a pleasure that has no 
annoying consequences, or one who avoids a pain that produces no resultant pleasure?" '
Annot: Table, text: ''
>>> 
```

**Reply by abhisarsaraswat** [reply]
Thanks for replying, Solution is excellent but, I can see in output that it only fetching text outputs for paragraphs only and not for headings.

**Reply by JorjMcKie** [reply]
this is because the annot for the heading has a rectangle which is too small to cover all of the underlying text:
![grafik](https://user-images.githubusercontent.com/8290722/126763334-c82c8dd8-a266-4ba4-b66c-375e088406e1.png)
Light yellow: the heading text, dark yellow the annot rect.
When extracting the text (`get_textbox()`), you could take a somewhat larger rectangle, like 5 points larger in every direction:
```python
>>> page.get_textbox(a.rect + (-5,-5,5,5))
'Lorem Ipsum - 1'
```

**Reply by JorjMcKie** [reply]
I don't know who created the annotations - but defining each of them large enough would of course do the same job.

**Reply by abhisarsaraswat** [reply]
can it be merged with logic you provided below?

> Would this be it:
> 
> ```python
> >>> for a in page.annots():
> 	print("Annot: %s, text: '%s'" % (a.info["content"], page.get_textbox(a.rect)))
> 
> 	
> Annot: Heading 1, text: ''
> Annot: Heading 2, text: ''
> Annot: Paragraph, text: '1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also
> the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing 
> Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum. '
> Annot: Reference Link, text: ''
> Annot: Heading 2, text: ''
> Annot: Paragraph, text: 'quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum 
> dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. '
> Annot: Heading 3, text: ''
> Annot: Paragraph, text: 'inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit, 
> sed quia consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt. Neque porro quisquam est, qui dolorem ipsum quia dolor sit amet, 
> consectetur, adipisci velit, sed quia non numquam eius modi tempora incidunt ut labore et dolore magnam aliquam quaerat voluptatem. Ut enim ad 
> minima veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem vel eum iure 
> reprehenderit qui in ea voluptate velit esse quam nihil molestiae consequatur, vel illum qui dolorem eum fugiat quo voluptas nulla pariatur? '
> Annot: Heading 4, text: ''
> Annot: Paragraph, text: 'system, and expound the actual teachings of the great explorer of the truth, the master-builder of human happiness. No one rejects, dislikes, or avoids 
> pleasure itself, because it is pleasure, but because those who do not know how to pursue pleasure rationally encounter consequences that are 
> extremely painful. Nor again is there anyone who loves or pursues or desires to obtain pain of itself, because it is pain, but because occasionally 
> circumstances occur in which toil and pain can procure him some great pleasure. To take a trivial example, which of us ever undertakes laborious 
> physical exercise, except to obtain some advantage from it? But who has any right to find fault with a man who chooses to enjoy a pleasure that has no 
> annoying consequences, or one who avoids a pain that produces no resultant pleasure?" '
> Annot: Table, text: ''
> >>> 
> ```

**Reply by JorjMcKie** [reply]
Example snippet:
```python
>>> doc=fitz.open("Segment_Data_Format.pdf")
>>> page=doc[0]
>>> a=page.first_annot
>>> page.delete_annot(a)
>>> rl=page.search_for("Lorem Ipsum - 1")
>>> rl
[Rect(72.0, 72.97378540039062, 244.26832580566406, 99.50157165527344)]
>>> a=page.add_highlight_annot(rl[0])
>>> page.get_textbox(a.rect)
'Lorem Ipsum - 1 '
>>> 
```

**Comment by JorjMcKie** [comment, marked as answer]
> can it be merged with logic you provided below?

Sure, why not

**Reply by JorjMcKie** [reply]
```python
for a in page.annots():
	print("Annot: %s, text: '%s'" % (a.info["content"], page.get_textbox(a.rect+(-5,-5,5,5))))
```

**Reply by abhisarsaraswat** [reply]
Thanks @JorjMcKie , You are a life saviour!
